File: packages/ascet-extension/ascet-cli/source/AscetTest/ContractValidator.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;

public sealed class AscetTestValidationIssue
{
    public string Path { get; set; }
    public string Code { get; set; }
    public string Message { get; set; }

    public Dictionary<string, object> ToDictionary()
    {
        return new Dictionary<string, object>
        {
            { "path", Path ?? String.Empty },
            { "code", Code ?? String.Empty },
            { "message", Message ?? String.Empty }
        };
    }
}

public sealed class AscetTestValidationResult
{
    public AscetTestValidationResult()
    {
        Errors = new List<AscetTestValidationIssue>();
        Warnings = new List<AscetTestValidationIssue>();
    }

    public IList<AscetTestValidationIssue> Errors { get; private set; }
    public IList<AscetTestValidationIssue> Warnings { get; private set; }
    public bool IsValid { get { return Errors.Count == 0; } }
}

public static class AscetTestContractValidator
{
    public const string RequestSchema = "ascet-test-request/v1";
    public const string ContractSchema = "ascet-test-contract/v1";

    public static AscetTestValidationResult ValidateRequest(
        IDictionary<string, object> request,
        IDictionary<string, object> contract)
    {
        AscetTestValidationResult result = new AscetTestValidationResult();
        if (request == null)
        {
            AddError(result, "$", "request_shape", "Request root must be an object.");
            return result;
        }

        RequireString(request, "schemaVersion", "schema_version", RequestSchema, result);
        RequireString(request, "runId", "required", null, result);
        ValidateRelativePath(GetString(request, "componentPath"), "componentPath", result);
        ValidateLevels(GetList(request, "levels"), "levels", result);

        bool executeLive = GetBoolean(request, "executeLive");
        if (executeLive && String.IsNullOrWhiteSpace(GetString(request, "approvedRevisionId")))
        {
            AddError(result, "approvedRevisionId", "approval_missing", "executeLive requires approvedRevisionId.");
        }

        if (contract == null)
        {
            AddError(result, "contract", "contract_missing", "Provide contract or testContractPath.");
        }
        else
        {
            string loadError = GetString(contract, "_loadError");
            if (!String.IsNullOrWhiteSpace(loadError))
            {
                AddError(result, "testContractPath", "test_contract_missing", loadError);
            }
            Merge(result, ValidateContract(contract));
            string requestComponent = NormalizePath(GetString(request, "componentPath"));
            string contractComponent = NormalizePath(GetString(contract, "componentPath"));
            if (!String.IsNullOrWhiteSpace(requestComponent) &&
                !String.IsNullOrWhiteSpace(contractComponent) &&
                !String.Equals(requestComponent, contractComponent, StringComparison.OrdinalIgnoreCase))
            {
                AddError(result, "contract.componentPath", "component_mismatch", "Contract componentPath must match request componentPath.");
            }
        }

        ValidateOptionalRunDirectory(request, result);
        return result;
    }

    public static AscetTestValidationResult ValidateContract(IDictionary<string, object> contract)
    {
        AscetTestValidationResult result = new AscetTestValidationResult();
        if (contract == null)
        {
            AddError(result, "contract", "contract_shape", "Contract must be an object.");
            return result;
        }

        if (!String.IsNullOrWhiteSpace(GetString(contract, "_loadError")))
        {
            return result;
        }

        RequireString(contract, "schemaVersion", "schema_version", ContractSchema, result);
        ValidateRelativePath(GetString(contract, "componentPath"), "contract.componentPath", result);
        IList suites = GetList(contract, "suites");
        if (suites == null || suites.Count == 0)
        {
            AddError(result, "contract.suites", "suites_missing", "Contract must contain at least one suite.");
            return result;
        }

        HashSet<string> suiteIds = new HashSet<string>(StringComparer.Ordinal);
        for (int index = 0; index < suites.Count; index++)
        {
            IDictionary<string, object> suite = AsDictionary(suites[index]);
            string path = "contract.suites[" + index.ToString() + "]";
            if (suite == null)
            {
                AddError(result, path, "suite_shape", "Suite must be an object.");
                continue;
            }

            string id = RequireString(suite, "id", "required", null, result, path + ".id");
            if (!String.IsNullOrWhiteSpace(id) && !suiteIds.Add(id))
            {
                AddError(result, path + ".id", "duplicate_suite", "Duplicate suite id '" + id + "'.");
            }

            string level = GetString(suite, "level");
            if (!IsLevel(level))
            {
                AddError(result, path + ".level", "invalid_level", "Suite level must be class_ut or component_ct.");
            }

            string entryPoint = RequireString(suite, "entryPoint", "required", null, result, path + ".entryPoint");
            string dependencyMode = GetString(suite, "dependencyMode");
            if (!String.Equals(dependencyMode, "stub", StringComparison.Ordinal) &&
                !String.Equals(dependencyMode, "real", StringComparison.Ordinal))
            {
                AddError(result, path + ".dependencyMode", "invalid_dependency_mode", "dependencyMode must be stub or real.");
            }
            if (String.Equals(level, "class_ut", StringComparison.Ordinal) && !String.Equals(dependencyMode, "stub", StringComparison.Ordinal))
            {
                AddError(result, path + ".dependencyMode", "class_ut_dependency_mode", "class_ut requires stub dependencies.");
            }
            if (String.Equals(level, "component_ct", StringComparison.Ordinal) && !String.Equals(dependencyMode, "real", StringComparison.Ordinal))
            {
                AddError(result, path + ".dependencyMode", "component_ct_dependency_mode", "component_ct requires real dependencies.");
            }

            int cycles = GetPositiveInteger(suite, "cycles", 1);
            if (cycles < 1)
            {
                AddError(result, path + ".cycles", "invalid_cycles", "cycles must be a positive integer.");
            }

            IList cases = GetList(suite, "cases");
            if (cases == null || cases.Count == 0)
            {
                AddError(result, path + ".cases", "cases_missing", "Suite must contain at least one case.");
                continue;
            }

            HashSet<string> caseIds = new HashSet<string>(StringComparer.Ordinal);
            for (int caseIndex = 0; caseIndex < cases.Count; caseIndex++)
            {
                IDictionary<string, object> testCase = AsDictionary(cases[caseIndex]);
                string casePath = path + ".cases[" + caseIndex.ToString() + "]";
                if (testCase == null)
                {
                    AddError(result, casePath, "case_shape", "Case must be an object.");
                    continue;
                }

                string caseId = RequireString(testCase, "id", "required", null, result, casePath + ".id");
                if (!String.IsNullOrWhiteSpace(caseId) && !caseIds.Add(caseId))
                {
                    AddError(result, casePath + ".id", "duplicate_case", "Duplicate case id '" + caseId + "'.");
                }

                if (GetPositiveInteger(testCase, "steps", 1) < 1)
                {
                    AddError(result, casePath + ".steps", "invalid_steps", "steps must be a positive integer.");
                }

                IDictionary<string, object> expectedOutputs = AsDictionary(GetValue(testCase, "expectedOutputs"));
                IList expectedSteps = GetList(testCase, "expectedSteps");
                if ((expectedOutputs == null || expectedOutputs.Count == 0) && (expectedSteps == null || expectedSteps.Count == 0))
                {
                    AddWarning(result, casePath, "trace_only_case", "Case has no expected outputs; it cannot prove a passing behavior.");
                }

                string oracleSource = GetString(testCase, "oracleSource");
                if (String.IsNullOrWhiteSpace(oracleSource))
                {
                    oracleSource = "derived";
                }
                if (!String.Equals(oracleSource, "requirement", StringComparison.Ordinal) &&
                    !String.Equals(oracleSource, "derived", StringComparison.Ordinal) &&
                    !String.Equals(oracleSource, "trace", StringComparison.Ordinal))
                {
                    AddError(result, casePath + ".oracleSource", "invalid_oracle_source", "oracleSource must be requirement, derived or trace.");
                }
            }
        }

        return result;
    }

    public static Dictionary<string, object> IssuesToPayload(IList<AscetTestValidationIssue> issues)
    {
        List<Dictionary<string, object>> payload = new List<Dictionary<string, object>>();
        if (issues != null)
        {
            for (int index = 0; index < issues.Count; index++)
            {
                payload.Add(issues[index].ToDictionary());
            }
        }
        return new Dictionary<string, object> { { "items", payload }, { "count", payload.Count } };
    }

    private static void ValidateLevels(IList levels, string path, AscetTestValidationResult result)
    {
        if (levels == null || levels.Count == 0)
        {
            AddError(result, path, "levels_missing", "levels must contain at least one level.");
            return;
        }

        HashSet<string> seen = new HashSet<string>(StringComparer.Ordinal);
        for (int index = 0; index < levels.Count; index++)
        {
            string level = Convert.ToString(levels[index]) ?? String.Empty;
            if (!IsLevel(level))
            {
                AddError(result, path + "[" + index.ToString() + "]", "invalid_level", "Level must be class_ut or component_ct.");
            }
            else if (!seen.Add(level))
            {
                AddError(result, path + "[" + index.ToString() + "]", "duplicate_level", "Duplicate level '" + level + "'.");
            }
        }
    }

    private static void ValidateOptionalRunDirectory(IDictionary<string, object> request, AscetTestValidationResult result)
    {
        string runDirectory = GetString(request, "runDirectory");
        if (String.IsNullOrWhiteSpace(runDirectory))
        {
            return;
        }

        if (runDirectory.IndexOf("..", StringComparison.Ordinal) >= 0)
        {
            AddError(result, "runDirectory", "unsafe_path", "runDirectory must not contain '..'.");
        }
    }

    private static bool IsLevel(string level)
    {
        return String.Equals(level, "class_ut", StringComparison.Ordinal) ||
               String.Equals(level, "component_ct", StringComparison.Ordinal);
    }

    private static string ValidateRelativePath(string value, string path, AscetTestValidationResult result)
    {
        if (String.IsNullOrWhiteSpace(value))
        {
            AddError(result, path, "required", path + " is required.");
            return String.Empty;
        }

        string normalized = NormalizePath(value);
        if (Path.IsPathRooted(value) || normalized.StartsWith("/", StringComparison.Ordinal) || normalized.Contains("../"))
        {
            AddError(result, path, "unsafe_path", path + " must be a relative ASCET path without '..'.");
        }
        return normalized;
    }

    private static string NormalizePath(string value)
    {
        return (value ?? String.Empty).Trim().Replace('\\', '/');
    }

    private static string RequireString(
        IDictionary<string, object> source,
        string key,
        string code,
        string expected,
        AscetTestValidationResult result,
        string path = null)
    {
        string value = GetString(source, key);
        string issuePath = String.IsNullOrWhiteSpace(path) ? key : path;
        if (String.IsNullOrWhiteSpace(value))
        {
            AddError(result, issuePath, "required", issuePath + " is required.");
        }
        else if (!String.IsNullOrWhiteSpace(expected) && !String.Equals(value, expected, StringComparison.Ordinal))
        {
            AddError(result, issuePath, code, issuePath + " must be " + expected + ".");
        }
        return value.Trim();
    }

    private static int GetPositiveInteger(IDictionary<string, object> source, string key, int defaultValue)
    {
        object value = GetValue(source, key);
        if (value == null)
        {
            return defaultValue;
        }

        try
        {
            int parsed = Convert.ToInt32(value);
            return parsed;
        }
        catch
        {
            return 0;
        }
    }

    private static bool GetBoolean(IDictionary<string, object> source, string key)
    {
        object value = GetValue(source, key);
        return value is bool && (bool)value;
    }

    private static string GetString(IDictionary<string, object> source, string key)
    {
        object value = GetValue(source, key);
        return value == null ? String.Empty : Convert.ToString(value) ?? String.Empty;
    }

    private static object GetValue(IDictionary<string, object> source, string key)
    {
        object value;
        return source != null && source.TryGetValue(key, out value) ? value : null;
    }

    private static IList GetList(IDictionary<string, object> source, string key)
    {
        object value = GetValue(source, key);
        return value as IList;
    }

    private static IDictionary<string, object> AsDictionary(object value)
    {
        return value as IDictionary<string, object>;
    }

    private static void Merge(AscetTestValidationResult target, AscetTestValidationResult source)
    {
        if (source == null)
        {
            return;
        }
        for (int index = 0; index < source.Errors.Count; index++) target.Errors.Add(source.Errors[index]);
        for (int index = 0; index < source.Warnings.Count; index++) target.Warnings.Add(source.Warnings[index]);
    }

    private static void AddError(AscetTestValidationResult result, string path, string code, string message)
    {
        result.Errors.Add(new AscetTestValidationIssue { Path = path, Code = code, Message = message });
    }

    private static void AddWarning(AscetTestValidationResult result, string path, string code, string message)
    {
        result.Warnings.Add(new AscetTestValidationIssue { Path = path, Code = code, Message = message });
    }
}

File: packages/ascet-extension/src/testing/contract-validator.ts
import {
	ASCET_TEST_CONTRACT_SCHEMA,
	ASCET_TEST_REQUEST_SCHEMA,
	type AscetTestCaseContract,
	type AscetTestContract,
	type AscetTestLevel,
	type AscetTestOracleSource,
	type AscetTestRequestDocument,
	type AscetTestSuiteContract,
	type AscetValidationIssue,
	type AscetValidationResult,
} from "./contracts.ts";

export function validateAscetTestRequest(value: unknown): AscetValidationResult<AscetTestRequestDocument> {
	const errors: AscetValidationIssue[] = [];
	const warnings: AscetValidationIssue[] = [];
	const record = asRecord(value);
	if (!record) {
		return { errors: [issue("$", "request_shape", "Request root must be an object.")], warnings };
	}

	if (record.schemaVersion !== ASCET_TEST_REQUEST_SCHEMA) {
		errors.push(issue("schemaVersion", "schema_version", `schemaVersion must be ${ASCET_TEST_REQUEST_SCHEMA}.`));
	}
	const runId = requiredString(record.runId, "runId", errors);
	const componentPath = validateRelativePath(record.componentPath, "componentPath", errors);
	const levels = validateLevels(record.levels, errors);
	const contract =
		record.contract === undefined ? undefined : validateContract(record.contract, errors, warnings).value;
	if (record.contract === undefined && !isNonEmptyString(record.testContractPath)) {
		errors.push(issue("contract", "contract_missing", "Provide contract or testContractPath."));
	}
	if (record.executeLive === true && !isNonEmptyString(record.approvedRevisionId)) {
		errors.push(issue("approvedRevisionId", "approval_missing", "executeLive requires approvedRevisionId."));
	}

	if (errors.length > 0) {
		return { errors, warnings };
	}
	return {
		errors,
		warnings,
		value: {
			schemaVersion: ASCET_TEST_REQUEST_SCHEMA,
			runId: runId!,
			componentPath: componentPath!,
			levels: levels!,
			contract,
			testContractPath: isNonEmptyString(record.testContractPath) ? record.testContractPath.trim() : undefined,
			runDirectory: isNonEmptyString(record.runDirectory) ? record.runDirectory.trim() : undefined,
			toolchain: asRecord(record.toolchain) as AscetTestRequestDocument["toolchain"],
			executeLive: record.executeLive === true,
			approvedRevisionId: isNonEmptyString(record.approvedRevisionId) ? record.approvedRevisionId.trim() : undefined,
		},
	};
}

export function validateAscetTestContract(value: unknown): AscetValidationResult<AscetTestContract> {
	const errors: AscetValidationIssue[] = [];
	const warnings: AscetValidationIssue[] = [];
	return validateContract(value, errors, warnings);
}

function validateContract(
	value: unknown,
	errors: AscetValidationIssue[],
	warnings: AscetValidationIssue[],
): AscetValidationResult<AscetTestContract> {
	const record = asRecord(value);
	if (!record) {
		errors.push(issue("contract", "contract_shape", "Contract must be an object."));
		return { errors, warnings };
	}
	if (record.schemaVersion !== ASCET_TEST_CONTRACT_SCHEMA) {
		errors.push(
			issue("contract.schemaVersion", "schema_version", `schemaVersion must be ${ASCET_TEST_CONTRACT_SCHEMA}.`),
		);
	}
	const componentPath = validateRelativePath(record.componentPath, "contract.componentPath", errors);
	const suitesRaw = Array.isArray(record.suites) ? record.suites : undefined;
	if (!suitesRaw || suitesRaw.length === 0) {
		errors.push(issue("contract.suites", "suites_missing", "Contract must contain at least one suite."));
		return { errors, warnings };
	}
	const suites: AscetTestSuiteContract[] = [];
	const suiteIds = new Set<string>();
	for (let index = 0; index < suitesRaw.length; index += 1) {
		const suite = validateSuite(suitesRaw[index], `contract.suites[${index}]`, errors, warnings);
		if (!suite) continue;
		if (suiteIds.has(suite.id)) {
			errors.push(issue(`contract.suites[${index}].id`, "duplicate_suite", `Duplicate suite id '${suite.id}'.`));
		}
		suiteIds.add(suite.id);
		suites.push(suite);
	}
	if (errors.length > 0) return { errors, warnings };
	return {
		errors,
		warnings,
		value: { schemaVersion: ASCET_TEST_CONTRACT_SCHEMA, componentPath: componentPath!, suites },
	};
}

function validateSuite(
	value: unknown,
	path: string,
	errors: AscetValidationIssue[],
	warnings: AscetValidationIssue[],
): AscetTestSuiteContract | undefined {
	const record = asRecord(value);
	if (!record) {
		errors.push(issue(path, "suite_shape", "Suite must be an object."));
		return undefined;
	}
	const id = requiredString(record.id, `${path}.id`, errors);
	const entryPoint = requiredString(record.entryPoint, `${path}.entryPoint`, errors);
	const level = record.level;
	if (!isTestLevel(level)) {
		errors.push(issue(`${path}.level`, "invalid_level", "Suite level must be class_ut or component_ct."));
	}
	const dependencyMode = record.dependencyMode;
	if (dependencyMode !== "stub" && dependencyMode !== "real") {
		errors.push(issue(`${path}.dependencyMode`, "invalid_dependency_mode", "dependencyMode must be stub or real."));
	}
	const cycles = record.cycles === undefined ? 1 : asPositiveInteger(record.cycles);
	if (cycles === undefined) {
		errors.push(issue(`${path}.cycles`, "invalid_cycles", "cycles must be a positive integer."));
	}
	if (level === "class_ut" && dependencyMode !== "stub") {
		errors.push(issue(`${path}.dependencyMode`, "class_ut_dependency_mode", "class_ut requires stub dependencies."));
	}
	if (level === "component_ct" && dependencyMode !== "real") {
		errors.push(
			issue(`${path}.dependencyMode`, "component_ct_dependency_mode", "component_ct requires real dependencies."),
		);
	}
	const casesRaw = Array.isArray(record.cases) ? record.cases : undefined;
	if (!casesRaw || casesRaw.length === 0) {
		errors.push(issue(`${path}.cases`, "cases_missing", "Suite must contain at least one case."));
		return undefined;
	}
	const cases: AscetTestCaseContract[] = [];
	const caseIds = new Set<string>();
	for (let index = 0; index < casesRaw.length; index += 1) {
		const current = validateCase(casesRaw[index], `${path}.cases[${index}]`, errors, warnings);
		if (!current) continue;
		if (caseIds.has(current.id)) {
			errors.push(issue(`${path}.cases[${index}].id`, "duplicate_case", `Duplicate case id '${current.id}'.`));
		}
		caseIds.add(current.id);
		cases.push(current);
	}
	if (!id || !entryPoint || !isTestLevel(level) || (dependencyMode !== "stub" && dependencyMode !== "real"))
		return undefined;
	return { id, level, entryPoint, dependencyMode, cycles, cases };
}

function validateCase(
	value: unknown,
	path: string,
	errors: AscetValidationIssue[],
	warnings: AscetValidationIssue[],
): AscetTestCaseContract | undefined {
	const record = asRecord(value);
	if (!record) {
		errors.push(issue(path, "case_shape", "Case must be an object."));
		return undefined;
	}
	const id = requiredString(record.id, `${path}.id`, errors);
	const steps = record.steps === undefined ? 1 : asPositiveInteger(record.steps);
	if (steps === undefined) errors.push(issue(`${path}.steps`, "invalid_steps", "steps must be a positive integer."));
	const expectedOutputs = asRecord(record.expectedOutputs);
	const expectedSteps = Array.isArray(record.expectedSteps) ? record.expectedSteps : [];
	if ((!expectedOutputs || Object.keys(expectedOutputs).length === 0) && expectedSteps.length === 0) {
		warnings.push(
			issue(path, "trace_only_case", "Case has no expected outputs; it is trace-only and cannot prove pass."),
		);
	}
	const oracleSourceValue = record.oracleSource === undefined ? "derived" : record.oracleSource;
	const oracleSource: AscetTestOracleSource | undefined = isOracleSource(oracleSourceValue)
		? oracleSourceValue
		: undefined;
	if (oracleSource === undefined) {
		errors.push(
			issue(`${path}.oracleSource`, "invalid_oracle_source", "oracleSource must be requirement, derived or trace."),
		);
	}
	if (!id || steps === undefined) return undefined;
	return {
		id,
		steps,
		inputs: asRecord(record.inputs) ?? {},
		expectedOutputs: expectedOutputs ?? {},
		expectedSteps: expectedSteps as AscetTestCaseContract["expectedSteps"],
		tolerances: asRecord(record.tolerances) as Record<string, number> | undefined,
		oracleSource,
		tags: Array.isArray(record.tags) ? record.tags.filter((tag): tag is string => typeof tag === "string") : [],
	};
}

function validateLevels(value: unknown, errors: AscetValidationIssue[]): AscetTestLevel[] | undefined {
	if (!Array.isArray(value) || value.length === 0) {
		errors.push(issue("levels", "levels_missing", "levels must contain at least one level."));
		return undefined;
	}
	const levels: AscetTestLevel[] = [];
	for (let index = 0; index < value.length; index += 1) {
		if (!isTestLevel(value[index])) {
			errors.push(issue(`levels[${index}]`, "invalid_level", "Level must be class_ut or component_ct."));
			continue;
		}
		if (levels.includes(value[index]))
			errors.push(issue(`levels[${index}]`, "duplicate_level", `Duplicate level '${value[index]}'.`));
		levels.push(value[index]);
	}
	return levels;
}

function validateRelativePath(value: unknown, path: string, errors: AscetValidationIssue[]): string | undefined {
	if (!isNonEmptyString(value)) {
		errors.push(issue(path, "required", `${path} is required.`));
		return undefined;
	}
	const normalized = value.trim().replace(/\\/g, "/");
	if (/^[A-Za-z]:\//.test(normalized) || normalized.startsWith("/") || normalized.split("/").includes("..")) {
		errors.push(issue(path, "unsafe_path", `${path} must be a relative ASCET path without '..'.`));
		return undefined;
	}
	return normalized;
}

function requiredString(value: unknown, path: string, errors: AscetValidationIssue[]): string | undefined {
	if (!isNonEmptyString(value)) {
		errors.push(issue(path, "required", `${path} is required.`));
		return undefined;
	}
	return value.trim();
}

function asPositiveInteger(value: unknown): number | undefined {
	return typeof value === "number" && Number.isInteger(value) && value >= 1 ? value : undefined;
}

function isTestLevel(value: unknown): value is AscetTestLevel {
	return value === "class_ut" || value === "component_ct";
}

function isOracleSource(value: unknown): value is AscetTestOracleSource {
	return value === "requirement" || value === "derived" || value === "trace";
}

function isNonEmptyString(value: unknown): value is string {
	return typeof value === "string" && value.trim().length > 0;
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
	return value !== null && typeof value === "object" && !Array.isArray(value)
		? (value as Record<string, unknown>)
		: undefined;
}

function issue(path: string, code: string, message: string): AscetValidationIssue {
	return { path, code, message };
}
